Implement proper version comparison for dev, alpha, beta and rc releases

phw opened the pull request:
<!--
    Hello! Thanks for submitting a pull request to MusicBrainz Picard. We
    appreciate your time and interest in helping our project!

    Use this template to help us review your change. Not everything is required,
    depending on your change. Keep or delete what is relevant for your change.
    Remember that it helps us review if you give more helpful info for us to
    understand your change.

    Ensure that you've read through and followed the Contributing Guidelines, in
    [CONTRIBUTING.md](https://github.com/metabrainz/picard/blob/master/CONTRIBUTING.md).
-->

# Summary
The version identifier was previously compared alphabetically, which fails for e.g. "dev" < "alpha".

<!--
    Update the checkbox with an [x] for the type of contribution you are making.
-->

* This is a…
    * [x] Bug fix
    * [ ] Feature addition
    * [x] Refactoring
    * [ ] Minor / simple change (like a typo)
    * [ ] Other
* **Describe this change in 1-2 sentences**:

# Problem
With the recent addition of (again) supporting alpha, beta and rc releases the comparison of the version identifier broke. This would lead to warnings such as "Warning: config file 2.3.0.dev1 was created by a more recent version of Picard (current is 2.3.0.alpha1)".
<!--
    Anything that helps us understand why you are making this change goes here.
    What problem are you trying to fix? What does this change address?
-->

* JIRA ticket (_optional_): PICARD-XXX
<!--
    Please make sure you prefix your pull request title with 'PICARD-XXX' in order
    for our ticket tracker to link your pull request to the relevant ticket.
-->



# Solution
Implement proper version comparison by introducing a Version class (inherited from tuple), which implements proper comparison operators. Also moved the logic fro to and from string conversion there.

`picard.version_from_string` is now considered obsolete, but I did not yet remove it as it might be used by plugin code. `version_to_string` is still useful when one has only a version tuple.

<!--
    The details of your change. Talk about technical details, considerations, or
    other interesting points. If you have a lot to say, be more detailed in this
    section.
-->


# Action

<!--
    Other than merging your change, do you want / need us to do anything else
    with your change? This could include reviewing a specific part of your PR.
-->

samj1912 commented on picard/version.py:
```diff
@@ -0,0 +1,123 @@
+    pass
+
+
+class Version(tuple):
```
Why not use total_ordering to get the comparators?

samj1912 replied:
Why not use a named tuple? From typing.

phw replied:
> Why not use total_ordering to get the comparators?

This would be best, but doesn't work since all comparators are already implemented (inherited from `tuple`) and `total_ordering` only fills in the missing ones. So we have to override them.

samj1912 commented on picard/version.py:
```diff
@@ -0,0 +1,123 @@
+        'final': 4
+    }
+
+    def __new__(cls, major, minor, patch, identifier='final', revision=0):
```
This could be init if named tuple is used?

phw replied:
No, actually not. init does not work with tuples at all, since they are immutable after creation. So even with nameduple we would need to use `__new__` (but only for the identifier validation). The code would change to something like:

```
class Version(namedtuple('V', 'major minor patch identifier revision')):

    def __new__(cls, *args, **kwargs):
        instance = super(Version, cls).__new__(cls, *args, **kwargs)
        if instance.identifier not in cls.valid_identifiers():
            raise VersionError("Should be either 'final', 'dev', 'alpha', 'beta' or 'rc'")
        return instance

```

I currently don't see much value in using namedtuple here. It introduces another class in between without bringing much benefit. Having the keywords available could make the code more readable at some places (e.g. `version.minor` instead of `version[1]`). I'll experiment with it and sees if is worth the trouble.

samj1912 commented on picard/version.py:
```diff
@@ -0,0 +1,123 @@
+
+    @classmethod
+    def valid_identifiers(cls):
+        return cls._identifiers.keys()
```
This will return dictkeys, it might be better returning a list or set or a tuple?

phw replied:
Yes, or make this thing private or remove it entirely. I'll look into it.

samj1912 commented on picard/version.py:
```diff
@@ -0,0 +1,123 @@
+    @classmethod
+    def from_string(cls, version_str):
+        m = cls._version_re.search(version_str)
+        if m:
```
Can we have better names than m and G?

samj1912 commented on picard/version.py:
```diff
@@ -0,0 +1,123 @@
+        m = cls._version_re.search(version_str)
+        if m:
+            g = m.groups()
+            if g[2] is None:
```
If named tuple is used, you can have defaults here.

phw commented:
@samj1912 I addressed your comments, including the use of namedtuple. We don't benefit much of the named fields right now, but it allows the code in `Version.to_string` to be a bit more readable.

zas commented on test/test_versions.py:
```diff
@@ -87,3 +98,54 @@ def test_api_versions_tuple_1(self):
+        v1 = Version(2, 3, 0, 'dev', 1)
+        v2 = Version(2, 3, 0, 'alpha', 1)
+        self.assertLess(v1, v2)
+        self.assertFalse(v2 < v2)
```
What's the goal of this test?

phw replied:
Verifying the __lt__ implementation compares properly
